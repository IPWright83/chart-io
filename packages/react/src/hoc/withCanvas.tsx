import React, { useEffect, useRef, useState } from "react";

import { useSelector } from "react-redux";
import { chartSelectors, IStore } from "../store";

/**
 * Wraps a D3 layer to make it work as a Canvas component
 * @param  WrappedComponent     The D3 layer to render to the Canvas
 * @param  className            An optional class name to add to the DOM
 * @return                      The wrapped layer
 */
const withCanvas =
    <P extends object>(WrappedComponent: React.ComponentType<P>, className: string) =>
    /**
     * Wraps a component within a Canvas
     * @param  {...any}    options.props        The rest of the props
     * @return {ReactDOMComponent}              The wrapped layer
     */
    (props) => {
        const canvas = useRef(null);
        const [layer, setLayer] = useState({});

        const width = useSelector((s: IStore) => chartSelectors.dimensions.width(s));
        const height = useSelector((s: IStore) => chartSelectors.dimensions.height(s));

        useEffect(() => {
            setLayer({ current: document.createElement("custom") });
        }, []);

        return (
            <React.Fragment>
                <foreignObject width={width} height={height} style={styles.foreignObject}>
                    <canvas
                        className={`canvas ${className}`}
                        width={width}
                        height={height}
                        ref={canvas}
                        style={styles.canvas}
                    ></canvas>
                </foreignObject>
                <WrappedComponent {...props} layer={layer} canvas={canvas.current} />
            </React.Fragment>
        );
    };

const styles = {
    canvas: {
        position: "absolute" as const,
    },
    foreignObject: {
        pointerEvents: "none" as const,
    },
};

export { withCanvas };
