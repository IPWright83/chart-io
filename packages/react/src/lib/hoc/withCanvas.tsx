import React, { useEffect, useRef, useState } from "react";

import { useSelector } from "react-redux";
import { chartSelectors, IState } from "../store";

/**
 * Wraps a D3 layer to make it work as a Canvas component
 * @param  WrappedComponent     The D3 layer to render to the Canvas
 * @param  className            An optional class name to add to the DOM
 * @return                      The wrapped layer
 */
const withCanvas = <P extends object>(WrappedComponent: React.ComponentType<P>, className: string) => {
    if (!className) {
        console.warn("W003 - className is required when using the withCanvas higher order component");
    }

    /**
     * Wraps a component within a Canvas
     * @param  {...any}            props        The rest of the props
     * @return {ReactDOMComponent}              The wrapped layer
     */
    return (props) => {
        const canvas = useRef(null);
        const [layer, setLayer] = useState({});

        const width = useSelector((s: IState) => chartSelectors.dimensions.width(s));
        const height = useSelector((s: IState) => chartSelectors.dimensions.height(s));

        useEffect(() => {
            setLayer({ current: document.createElement("custom") });
        }, []);

        return (
            <React.Fragment>
                <foreignObject width={width} height={height} style={styles.foreignObject}>
                    <canvas
                        className={`canvas ${className}`}
                        width={width}
                        height={height}
                        ref={canvas}
                        style={styles.canvas}
                    ></canvas>
                </foreignObject>
                <WrappedComponent {...props} layer={layer} canvas={canvas.current} />
            </React.Fragment>
        );
    };
};

const styles = {
    canvas: {
        position: "absolute" as const,
    },
    foreignObject: {
        pointerEvents: "none" as const,
    },
};

export { withCanvas };
