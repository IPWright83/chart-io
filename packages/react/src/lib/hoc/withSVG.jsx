import React, { useRef } from "react";

/**
 * Wraps a D3 layer to make it work as an SVG component
 * @param  {ReactDOMComponent} WrappedComponent     The D3 layer to render to the SVG
 * @param  {String}            className            An optional class name to add to the DOM
 * @return {ReactDOMComponent}                      The wrapped layer
 */
const withSVG = (WrappedComponent, className) => {
    if (!className) {
        console.warn("W003 - className is required when using the withSVG higher order component");
    }

    /**
     * Wraps a component within an SVG group
     * @param  {...any}            props        The rest of the props
     * @return {ReactDOMComponent}              The wrapped layer
     */
    return (props) => {
        const layer = useRef(null);

        return (
            <g ref={layer} className={`g-${className}`}>
                <React.StrictMode>
                    <WrappedComponent {...props} layer={layer} />
                </React.StrictMode>
            </g>
        );
    };
};

export { withSVG };
