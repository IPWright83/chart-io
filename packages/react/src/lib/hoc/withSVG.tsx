import React, { useRef } from "react";
import { logWarning } from "../utils";

import { logWarning } from "../utils";

/**
 * Wraps a D3 layer to make it work as an SVG component
 * @param  WrappedComponent     The D3 layer to render to the SVG
 * @param  className            An optional class name to add to the DOM
 * @return                      The wrapped layer
 */
const withSVG = <P extends object>(WrappedComponent: React.ComponentType<P>, className: string) => {
    if (!className) {
        logWarning("W003", "className is required when using the withSVG higher order component");
    }

    /**
     * Wraps a component within an SVG group
     * @param  {...any}            props        The rest of the props
     * @return {ReactDOMComponent}              The wrapped layer
     */
    return function withSVG(props) {
        const layer = useRef(null);

        return (
            <g ref={layer} className={`g-${className}`}>
                <React.StrictMode>
                    <WrappedComponent {...props} layer={layer} />
                </React.StrictMode>
            </g>
        );
    };
};

export { withSVG };
