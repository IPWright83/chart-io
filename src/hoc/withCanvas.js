import React, { useEffect, useRef, useState } from "react";

import { useSelector } from "react-redux";
import { chartSelectors } from "../store";

/**
 * Wraps a D3 layer to make it work as a Canvas component
 * @param  {ReactDOMComponent} WrappedComponent     The D3 layer to render to the Canvas
 * @param  {String}            className            An optional class name to add to the DOM
 * @return {ReactDOMComponent}                      The wrapped layer
 */
const withCanvas = (WrappedComponent, className) =>
    /**
     * Wraps a component within a Canvas
     * @param  {...any}    options.props        The rest of the props
     * @return {ReactDOMComponent}              The wrapped layer
     */
    (props) => {
        const canvas = useRef(null);
        const [layer, setLayer] = useState({});

        const width = useSelector((s) => chartSelectors.dimensions.width(s));
        const height = useSelector((s) => chartSelectors.dimensions.height(s));

        useEffect(() => {
            const element = document.createElement("custom");
            document.body.appendChild(element);
            element.className = `g-${className}`;
            console.log(element.className);
            setLayer({ current: element });
        }, []);

        return (
            <React.Fragment>
                <foreignObject width={width} height={height} style={styles.foreignObject}>
                    <canvas width={width} height={height} ref={canvas} style={styles.canvas}></canvas>
                </foreignObject>
                <WrappedComponent {...props} layer={layer} canvas={canvas.current} />
            </React.Fragment>
        );
    };

const styles = {
    canvas: {
        position: "absolute",
    },
    foreignObject: {
        pointerEvents: "none",
    },
};

export { withCanvas };
