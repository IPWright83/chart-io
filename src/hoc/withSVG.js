import React, { useRef } from "react";

/**
 * Wraps a D3 layer to make it work as an SVG component
 * @param  {ReactDOMComponent} WrappedComponent     The D3 layer to render to the SVG
 * @return {ReactDOMComponent}                      The wrapped layer
 */
const withSVG = (WrappedComponent) =>
    /**
     * Wraps a component within an SVG group
     * @param  {...any}    options.props        The rest of the props
     * @return {ReactDOMComponent}              The wrapped layer
     */
    (props) => {
        const layer = useRef(null);

        return (
            <g ref={layer}>
                <WrappedComponent {...props} layer={layer} />
            </g>
        );
    };

export { withSVG };
