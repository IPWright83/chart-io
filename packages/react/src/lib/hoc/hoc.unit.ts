import React from "react";

import { withCanvas } from "./withCanvas";
import { withSVG } from "./withSVG";

describe("Higher Order Components", () => {
    describe("withCanvas", () => {
        it("should warn if no classname is provided", () => {
            const mockElement = React.createElement("div");

            const spy = jest.spyOn(console, "warn").mockImplementation();
            // @ts-expect-error: Validating a runtime warning
            withCanvas(mockElement);

            expect(spy.mock.calls[0][0]).toMatchSnapshot();
        });
    });

    describe("withSVG", () => {
        it("should warn if no classname is provided", () => {
            const mockElement = React.createElement("div");

            const spy = jest.spyOn(console, "warn").mockImplementation();
            // @ts-expect-error: Validating a runtime warning
            withSVG(mockElement);

            expect(spy.mock.calls[0][0]).toMatchSnapshot();
        });
    });
});
