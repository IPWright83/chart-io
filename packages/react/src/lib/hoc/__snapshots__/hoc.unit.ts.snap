// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Higher Order Components withCanvas should warn if no classname is provided 1`] = `"@chart-it/react encountered an warning. W003: className is required when using the withCanvas higher order component. You can read more about this https://ipwright83.github.io/chart-it/?path=/docs/errors-warnings-warnings-W003."`;

exports[`Higher Order Components withSVG should warn if no classname is provided 1`] = `"@chart-it/react encountered an warning. W003: className is required when using the withSVG higher order component. You can read more about this https://ipwright83.github.io/chart-it/?path=/docs/errors-warnings-warnings-W003."`;
